osquery/events/darwin/endpointsecurity_fim.cpp

```diff
@@ -0,0 +1,196 @@
+      return Status::failure(
+          1, "EndpointSecurity FIM is disabled via configuration");
+    }
+
+    if (!FLAGS_es_fim_mute_path_literal.empty()) {
+      boost::split(muted_path_literals_,
+                   FLAGS_es_fim_mute_path_literal,
+                   boost::is_any_of(","));
+      VLOG(1) << "muted path list :: " << muted_path_literals_.size();
+      for (const auto& p : muted_path_literals_) {
+        VLOG(1) << p;
+      }
+    }
+
+    if (!FLAGS_es_fim_mute_path_prefix.empty()) {
+      boost::split(muted_path_prefixes_,
+                   FLAGS_es_fim_mute_path_prefix,
+                   boost::is_any_of(","));
+      ;
+    }
+
+    auto handler = ^(es_client_t* client, const es_message_t* message) {
+      handleMessage(message);
+    };
+
+    auto result = es_new_client(&es_file_client_, handler);
+    if (result == ES_NEW_CLIENT_RESULT_SUCCESS) {
+      es_file_client_success_ = true;
+      return Status::success();
+    } else {
+      return Status::failure(1, getEsNewClientErrorMessage(result));
+    }
+  } else {
+    return Status::failure(
+        1, "EndpointSecurity is only available on macOS 10.15 and higher");
+  }
+}
+
+void EndpointSecurityFileEventPublisher::configure() {
+  if (es_file_client_ == nullptr) {
+    return;
+  }
+
+  auto cache = es_clear_cache(es_file_client_);
+  if (cache != ES_CLEAR_CACHE_RESULT_SUCCESS) {
+    VLOG(1) << "Couldn't clear cache for EndpointSecurity client";
+    return;
+  }
+
+  for (auto& sub : subscriptions_) {
+    auto sc = getSubscriptionContext(sub->context);
+    auto events = sc->es_file_event_subscriptions_;
+
+    for (const auto& p : muted_path_literals_) {
+      auto result = es_mute_path_literal(es_file_client_, p.c_str());
+      if (result == ES_RETURN_ERROR) {
+        VLOG(1) << "Unable to mute path literal: " << p;
+      }
+    }
+
+    for (const auto& p : muted_path_prefixes_) {
+      auto result = es_mute_path_prefix(es_file_client_, p.c_str());
+      if (result == ES_RETURN_ERROR) {
+        VLOG(1) << "Unable to mute path with prefix: " << p;
+      }
+    }
+    auto es_sub = es_subscribe(es_file_client_, &events[0], events.size());
+    if (es_sub != ES_RETURN_SUCCESS) {
+      VLOG(1) << "Couldn't subscribe to EndpointSecurity subsystem";
+    }
+  }
+}
+
+bool EndpointSecurityFileEventPublisher::shouldFire(
+    const EndpointSecurityFileSubscriptionContextRef& sc,
+    const EndpointSecurityFileEventContextRef& ec) const {
+  return true;
+}
+
+void EndpointSecurityFileEventPublisher::tearDown() {
+  if (es_file_client_ == nullptr) {
+    return;
+  }
+  es_unsubscribe_all(es_file_client_);
+  if (es_file_client_success_) {
+    es_delete_client(es_file_client_);
+    es_file_client_ = nullptr;
+  }
+}
+
+void EndpointSecurityFileEventPublisher::handleMessage(
+    const es_message_t* message) {
+  if (message == nullptr) {
+    return;
+  }
+
+  if (message->action_type == ES_ACTION_TYPE_AUTH) {
```

Comment:
This check is for sanity only? I don't see a case that you subscribe to notification event and retrieve authorization event.

Reply:
Yep, just for sanity and defensive coding